server/src/main/java/org/elasticsearch/ingest/FieldInferenceBulkRequestPreprocessor.java

```diff
@@ -0,0 +1,124 @@
+import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.client.internal.OriginSettingClient;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.inference.TaskType;
+import org.elasticsearch.plugins.internal.DocumentParsingObserver;
+
+import java.util.Map;
+import java.util.function.BiConsumer;
+import java.util.function.IntConsumer;
+import java.util.function.Supplier;
+
+public class FieldInferenceBulkRequestPreprocessor extends AbstractBulkRequestPreprocessor {
+
+    public static final String SEMANTIC_TEXT_ORIGIN = "semantic_text";
+
+    private final OriginSettingClient client;
+
+    public FieldInferenceBulkRequestPreprocessor(Supplier<DocumentParsingObserver> documentParsingObserver, Client client) {
+        super(documentParsingObserver);
+        this.client = new OriginSettingClient(client, SEMANTIC_TEXT_ORIGIN);
+    }
+
+    protected void processIndexRequest(
+        IndexRequest indexRequest,
+        int slot,
+        RefCountingRunnable refs,
+        IntConsumer onDropped,
+        final BiConsumer<Integer, Exception> onFailure
+    ) {
+        assert indexRequest.isFieldInferenceDone() == false;
+
+        String index = indexRequest.index();
+        Map<String, Object> sourceMap = indexRequest.sourceAsMap();
+        sourceMap.entrySet().stream().filter(entry -> fieldNeedsInference(index, entry.getKey())).forEach(entry -> {
+            runInferenceForField(indexRequest, entry.getKey(), refs, slot, onFailure);
+        });
+    }
+
+    @Override
+    public boolean needsProcessing(DocWriteRequest<?> docWriteRequest, IndexRequest indexRequest, Metadata metadata) {
+        return (indexRequest.isFieldInferenceDone() == false)
+            && indexRequest.sourceAsMap().keySet().stream().anyMatch(fieldName -> fieldNeedsInference(indexRequest.index(), fieldName));
+    }
+
+    @Override
+    public boolean hasBeenProcessed(IndexRequest indexRequest) {
+        return indexRequest.isFieldInferenceDone();
+    }
+
+    @Override
+    public boolean shouldExecuteOnIngestNode() {
+        return false;
+    }
+
+    // TODO actual mapping check here
+    private boolean fieldNeedsInference(String index, String fieldName) {
+        return fieldName.startsWith("infer_");
+    }
+
+    private void runInferenceForField(
+        IndexRequest indexRequest,
+        String fieldName,
+        RefCountingRunnable refs,
+        int position,
+        BiConsumer<Integer, Exception> onFailure
+    ) {
+        var ingestDocument = newIngestDocument(indexRequest);
+        if (ingestDocument.hasField(fieldName) == false) {
+            return;
+        }
+
+        refs.acquire();
+
+        // TODO Hardcoding model ID and task type
+        InferenceAction.Request inferenceRequest = new InferenceAction.Request(
+            TaskType.SPARSE_EMBEDDING,
+            "my-elser-model",
+            ingestDocument.getFieldValue(fieldName, String.class),
+            Map.of()
+        );
```

Comment:
This would come from the mapping IMO. The mapping should have the `model` information.

Reply:
💯 . We'll be adding that as part of a separate PoC; once the semantic_text field is added, then the fields that need inference will be added as part of IndexMetadata or MappingMetadata so we can access them efficiently.